plugins/telink-dfu/fu-telink-dfu-hid-device.c

```diff
@@ -0,0 +1,409 @@
+/*
+ * Copyright 2024 Mike Chang <Mike.chang@telink-semi.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ */
+
+#include "config.h"
+
+#include <linux/hidraw.h>
+#include <linux/input.h>
+
+#include "fu-telink-dfu-archive.h"
+#include "fu-telink-dfu-hid-device.h"
+#include "fu-telink-dfu-struct.h"
+
+struct _FuTelinkDfuHidDevice {
+	FuHidDevice parent_instance;
+	guint16 hid_tool_ver;
+};
+
+G_DEFINE_TYPE(FuTelinkDfuHidDevice, fu_telink_dfu_hid_device, FU_TYPE_HID_DEVICE)
+
+#define FU_TELINK_DFU_HID_DEVICE_START_ADDR	0x0000
+#define FU_TELINK_DFU_HID_DEVICE_REPORT_TIMEOUT 500 /* ms */
+#define FU_TELINK_DFU_HID_DEVICE_OTA_LENGTH	65
+#define FU_TELINK_DFU_HID_DEVICE_OTA_START_LEN	2
+#define FU_TELINK_DFU_HID_DEVICE_OTA_END_LEN	6
+#define FU_TELINK_DFU_HID_DEVICE_OTA_DATA_LEN	20
+#define FU_TELINK_DFU_HID_DEVICE_REPORT_ID	6
+#define HID_IFACE				2
+#define HID_EP_IN				(0x80 | 4)
+#define HID_EP_OUT				(0x00 | 5)
+#define HID_TOOL_VER(a, b)			((a) * 100 + (b))
+
+static gboolean
+fu_telink_dfu_hid_device_setup(FuDevice *device, GError **error)
+{
+	/* HidDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_telink_dfu_hid_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	/* success */
+	return TRUE;
+}
```

Comment:
do you need this section at all? It'll fall back to `HidDevice->setup` if not specified.